File: 100_Doors_Problem/C++/tylerferraro/100doors.cpp
#include <iostream>
#include <bitset>

#define MAX_DOORS 100

using std::cout;
using std::endl;
using std::bitset;

void print(bitset<MAX_DOORS> &doors)
{
    for (int i = 0; i < MAX_DOORS; i++)
    {
        if (doors[i] == 1)
            cout << "Door #" << i + 1 << ": " << "Open" << endl;
    }

    for (int i = 0; i < MAX_DOORS; i++)
    {
        if (doors[i] == 0)
            cout << "Door #" << i + 1 << ": " << "Closed" << endl;
    }
}

void walk_hallway(bitset<MAX_DOORS> &doors, int passes)
{
    for (int pass = 1; pass < passes + 1; pass++)
    {
        for (int i = pass - 1; i < MAX_DOORS; i += pass)
        {
            doors.flip(i);
        }
    }
}

int main()
{
    bitset<MAX_DOORS> doors;
    walk_hallway(doors, 100);
    print(doors);

    return 0;
}

File: 100_Doors_Problem/Ruby/tylerferraro/100doors.rb
#!/usr/bin/ruby

def print_hallway(doors)
  doors.each_with_index { |status, i| puts "Door \##{i+1}: Open" if status == 1 }
  doors.each_with_index { |status, i| puts "Door \##{i+1}: Closed" if status == 0 }
end

# Each pass start at the nth door, open/close it
# then step n more doors away and repeat
def walk_hallway(doors, passes)
  passes.times do |n|
    (n..passes).step(n+1) { |i| doors[i] ^= 1 }
  end
end

MAX_DOORS = 100
MAX_PASSES = 100

doors = [0]*MAX_DOORS
walk_hallway(doors, MAX_PASSES)
print_hallway(doors)
